### DISCUSSION TITLE
Injection with DI

### DISCUSSION THREAD

**Discussion by vasiliykai** [mganss/HtmlSanitizer#314] [General]
Hi! 
Injection with DI pass empty lists into constructor instead of expected null parameteres. Because of that default allowed tags not initialize. Problem can be solve in DI registration. But this this is unexpected behavior. 
Thank you!

**Comment by mganss** [comment]
Do you have an idea how this could be solved in a backwards compatible manner?

**Comment by tiesont** [comment]
I think you'd need to verify if this happens with all DI frameworks, or just specific versions. If it's everything, that would imply that is expected behavior.

**Reply by vasiliykai** [reply]
User of library no need to know about its realization. Allowed tags must works on default. And becouse of that anyway it is unexpected behavior

**Comment by vasiliykai** [comment]
You can change the constructor like that:
public HtmlSanitizer(...)
{
    AllowedTags = new HashSet<string>(string.IsNullOrEmpty(allowedTags.FirstOrDefault()) ? DefaultAllowedTags :  allowedTags , 
        StringComparer.OrdinalIgnoreCase);
    ...
}

**Reply by mganss** [reply]
> You can change the constructor like that: public HtmlSanitizer(...) { AllowedTags = new HashSet(string.IsNullOrEmpty(allowedTags.FirstOrDefault()) ? DefaultAllowedTags : allowedTags , StringComparer.OrdinalIgnoreCase); ... }

This would be a breaking change for people who pass an empty list to indicate that they want to disallow all tags. Also it would throw when passing `null`.

**Reply by mganss** [reply]
Perhaps we could treat the special case where all arguments are empty lists in the same as when they are all null.

**Reply by vasiliykai** [reply]
Also you can add default contructor without parameters and configure it to call original constructor with required parameters. DI always call default contructor by default

**Reply by mganss** [reply]
Are you sure about this? Do you have a link to documentation about this? AFAIK it will try to call the constructor having the most resolvable arguments.

**Reply by vasiliykai** [reply]
Yes, you right. Maybe all you can is to make breaking-change patch and make decisions about allowed tags more explicit. You can improve convenience by give users easily way to add their tags to default tags, as example... Or you can add method like ResetAllAllowsToDefautl(). I really dont know which way is better. If i have some ideas i wrote you

**Reply by mganss** [reply]
You can easily manipulate the default sets through `DefaultAllowedTags` etc.

**Comment by mganss** [comment]
See also #220.

**Comment by mganss** [comment]
I've fixed this by treating the special case where all arguments are empty the same as if they were all nulls.
